colexec: improve hash joiner benchmarks

yuzefovich opened the pull request:
This commit improves the hash joiner benchmarks in the following manner:
- it adds BYTES type (previously, we only used INT)
- it fully exhausts the hash joiner (previously, we would stop at an
arbitrary point, as if the query had a LIMIT; this could actually hide
some aspects of the performance - when Same slice is fully built)
- it makes it so that the data in `rightDistinct=true` scenario is
actually distinct
- it removes `fullOuter` dimension while only keeping the inner join
(this dimension doesn't seem that useful, relatively speaking)
- it adds another dimension for the number of equality columns
(previously, we would only run with 2 eq cols, now we run with 1 and 3)
- it adjusts `nulls=true` case to use NULL values randomly with 10%
probability (rather than only setting the first tuple to be all NULLs)
- it introduces another `leftDistinct` dimension.

Probably the biggest change is how the input tuples are populated.
Previously, we created a single batch with 1024 different rows, and that
batch would be repeated certain number of times on both sides. This
doesn't matter that much for the right source (because there we simply
buffer all tuples anyway), but it's not really reasonable for the left
source - every probing batch is the duplicate of the first batch. This
commit changes that so that we generate a set of tuples of the target
length in the ordered fashion, then shuffle the set, and chunk it up
into separate batches. This is likely to be a better approximation of
the production data sets - each probing batch might now have some new
tuples never seen previously.

All of these changes make the benchmark not comparable to before this
commit, so this commit will be backported to 23.1.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101974)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: improve hash joiner benchmarks

This commit improves the hash joiner benchmarks in the following manner:
- it adds BYTES type (previously, we only used INT)
- it fully exhausts the hash joiner (previously, we would stop at an
arbitrary point, as if the query had a LIMIT; this could actually hide
some aspects of the performance - when Same slice is fully built)
- it makes it so that the data in `rightDistinct=true` scenario is
actually distinct
- it removes `fullOuter` dimension while only keeping the inner join
(this dimension doesn't seem that useful, relatively speaking)
- it adds another dimension for the number of equality columns
(previously, we would only run with 2 eq cols, now we run with 1 and 3)
- it adjusts `nulls=true` case to use NULL values randomly with 10%
probability (rather than only setting the first tuple to be all NULLs)
- it introduces another `leftDistinct` dimension.

Probably the biggest change is how the input tuples are populated.
Previously, we created a single batch with 1024 different rows, and that
batch would be repeated certain number of times on both sides. This
doesn't matter that much for the right source (because there we simply
buffer all tuples anyway), but it's not really reasonable for the left
source - every probing batch is the duplicate of the first batch. This
commit changes that so that we generate a set of tuples of the target
length in the ordered fashion, then shuffle the set, and chunk it up
into separate batches. This is likely to be a better approximation of
the production data sets - each probing batch might now have some new
tuples never seen previously.

All of these changes make the benchmark not comparable to before this
commit, so this commit will be backported to 23.1.

Epic: None

Release note: None

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice. Just nits.

Would it make sense to factor a common helper function out of both `BenchmarkExternalHashJoiner` and `BenchmarkHashJoiner` with these changes? It seems like both benchmarks could benefit from these improvements.

> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101974#-:-NU3ppwJEZueKvoAiL3r:b-5winqx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/colexec/external_hash_joiner_test.go` line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101974#-NU3fD308I7WRTznww3M:-NU3fD308I7WRTznww3N:b-3azz1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ff6a1325cd48eb1fbad0497aca6c5e91197b0da/pkg/sql/colexec/external_hash_joiner_test.go#L191)):*
> ```Go
> }
> 
> // newBytesColumns returns nCols columns of types.Int with non-decreasing values
> ```

nit: "types.Int" should be "types.Bytes"
___
*[`pkg/sql/colexec/external_hash_joiner_test.go` line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101974#-NU3mkDc215xnIq1W-tY:-NU3mkDc215xnIq1W-tZ:b-462o90) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ff6a1325cd48eb1fbad0497aca6c5e91197b0da/pkg/sql/colexec/external_hash_joiner_test.go#L192)):*
> ```Go
> 
> // newBytesColumns returns nCols columns of types.Int with non-decreasing values
> // of 8 byte size, starting at '00000000'. dupCount controls the number of
> ```

Not a request, just a comment: I think width of the equality columns would also be an interesting dimension. (Though maybe some of that is covered by the number-of-equality-columns dimension.)
___
*[`pkg/sql/colexec/hashjoiner_test.go` line 1122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101974#-NU3pMjh4x4VYgULV71k:-NU3pMjh4x4VYgULV71l:bp58lvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ff6a1325cd48eb1fbad0497aca6c5e91197b0da/pkg/sql/colexec/hashjoiner_test.go#L1122)):*
> ```Go
> 							eqCols := []uint32{0, 1, 2}[:nEqCols]
> 							hjSpec := colexecjoin.MakeHashJoinerSpec(
> 								descpb.InnerJoin, eqCols, eqCols,
> ```

Would inner join vs full outer join also be a dimension worth measuring, as is done in `BenchmarkExternalHashJoiner`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I actually somewhat intentionally kept `BenchmarkExternalHashJoiner` mostly unchanged - i.e. preserved the same data set construction as well as kept the fullOuter dimension. Now two benchmarks provide us somewhat unique signals which seems more helpful than having a very similar setup in both making two signals somewhat redundant.

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101974#-:-NU4EQqs2SVpSdlzp5UN:bpbh0ij)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/colexec/external_hash_joiner_test.go` line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101974#-NU3mkDc215xnIq1W-tY:-NU4D42H7Ew9qScG62az:btna1jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ff6a1325cd48eb1fbad0497aca6c5e91197b0da/pkg/sql/colexec/external_hash_joiner_test.go#L192)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Not a request, just a comment: I think width of the equality columns would also be an interesting dimension. (Though maybe some of that is covered by the number-of-equality-columns dimension.)
</blockquote></details>

I agree that'd be interesting. I thought about this too, but the benchmark is already quite slow, and an extra dimension would only slow down the iteration speed (on making changes to the hash joiner and measuring the impact).

Different widths here would change the cost of individual comparisons, and to a certain degree that was the reason for why I introduced Bytes type in this commit - I have some improvements brewing that were not visible with just ints, with bytes the changes are more pronounced. If we find another scenario where expected improvements don't show up in the microbenchmarks because the data sets aren't interesting enough, then we'll add another dimension.
___
*[`pkg/sql/colexec/hashjoiner_test.go` line 1122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101974#-NU3pMjh4x4VYgULV71k:-NU4E9O12VBvSGQwHDhy:b-3toviq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ff6a1325cd48eb1fbad0497aca6c5e91197b0da/pkg/sql/colexec/hashjoiner_test.go#L1122)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would inner join vs full outer join also be a dimension worth measuring, as is done in `BenchmarkExternalHashJoiner`?
</blockquote></details>

I found it not very interesting and decided to remove it to shorten the benchmark runtime (it's still pretty long, but I couldn't find another dimension that would be not very useful; perhaps `nulls`, but that seemed more interesting that `fullOuter` dimension).


<!-- Sent from Reviewable.io -->
